tests/e2e/json_utils.go

```diff
@@ -0,0 +1,412 @@
+		}
+	case "main.StartSovereignChainAction":
+		var a StartSovereignChainAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.lightClientEquivocationAttackAction":
+		var a lightClientEquivocationAttackAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.lightClientAmnesiaAttackAction":
+		var a lightClientAmnesiaAttackAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.lightClientLunaticAttackAction":
+		var a lightClientLunaticAttackAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	default:
+		return nil, fmt.Errorf("unknown action name: %s", actionTypeString)
+	}
+	return nil, err
+}
+
+// custom marshal and unmarshal functions for the chainstate that convert proposals to/from the auxiliary type with type info
+
+// transform the ChainState into a ChainStateWithProposalTypes by adding type info to the proposals
+func (c ChainState) MarshalJSON() ([]byte, error) {
+	type ProposalAndType struct {
+		RawProposal interface{}
+		Type        string
+	}
+
+	type ChainStateWithProposalTypes struct {
+		ValBalances                    *map[ValidatorID]uint
+		ValPowers                      *map[ValidatorID]uint
+		RepresentativePowers           *map[ValidatorID]uint
+		Params                         *[]Param
+		Rewards                        *Rewards
+		ConsumerChains                 *map[ChainID]bool
+		AssignedKeys                   *map[ValidatorID]string
+		ProviderKeys                   *map[ValidatorID]string
+		ConsumerChainQueueSizes        *map[ChainID]uint
+		GlobalSlashQueueSize           *uint
+		RegisteredConsumerRewardDenoms *[]string
+		Proposals                      *map[uint]ProposalAndType // the only thing changed from the real ChainState
+	}
+
+	chainStateWithProposalTypes := ChainStateWithProposalTypes{
+		ValBalances:                    c.ValBalances,
+		ValPowers:                      c.ValPowers,
+		RepresentativePowers:           c.RepresentativePowers,
+		Params:                         c.Params,
+		Rewards:                        c.Rewards,
+		ConsumerChains:                 c.ConsumerChains,
+		AssignedKeys:                   c.AssignedKeys,
+		ProviderKeys:                   c.ProviderKeys,
+		ConsumerChainQueueSizes:        c.ConsumerChainQueueSizes,
+		GlobalSlashQueueSize:           c.GlobalSlashQueueSize,
+		RegisteredConsumerRewardDenoms: c.RegisteredConsumerRewardDenoms,
+	}
+	if c.Proposals != nil {
+		proposalsWithTypes := make(map[uint]ProposalAndType)
+		for k, v := range *c.Proposals {
+			proposalsWithTypes[k] = ProposalAndType{v, reflect.TypeOf(v).String()}
+		}
+		chainStateWithProposalTypes.Proposals = &proposalsWithTypes
+	}
+	return json.Marshal(chainStateWithProposalTypes)
+}
+
+// unmarshal the ChainStateWithProposalTypes into a ChainState by removing the type info from the proposals and getting back standard proposals
```

Comment:
Use the name of the method as the first thing in the comment for godocs? Similarly above.

Reply:
Thanks, good suggestion. Done, d1465053a4495a90af3a12c92842ecb2e3a5ec21